apps/browser-extension-wallet/src/views/browser-view/features/swaps/components/SwapProvider.tsx

```diff
@@ -0,0 +1,442 @@
+      const postBody = JSON.stringify(
+        createSwapRequestBody({
+          tokenA: tokenA.id,
+          tokenB: tokenB.policyId + tokenB.policyName,
+          quantity,
+          ignoredDexs: excludedDexs,
+          address: addresses?.[0]?.address,
+          targetSlippage,
+          collateral,
+          utxos
+        })
+      );
+
+      const response = await window.fetch(`${process.env.STEELSWAP_API_URL}/swap/build/`, {
+        method: 'POST',
+        headers: createSteelswapApiHeaders(),
+        body: postBody
+      });
+      const unableToBuildErrorText = t('swaps.error.unableToBuild');
+      if (!response.ok) {
+        try {
+          const { detail } = await response.json();
+          // 406 status indicates a specific error that should be shown to user
+          if (response.status === HttpStatusCode.NotAcceptable) {
+            toast.notify({ duration: 3, text: detail });
+            return;
+          }
+          // For other error statuses, show generic error and log details
+          logger.error('Failed to build swap:', { status: response.status, detail });
+          toast.notify({ duration: 3, text: unableToBuildErrorText });
+          return;
+        } catch {
+          logger.error('Failed to build swap: unable to parse error response');
+          toast.notify({ duration: 3, text: unableToBuildErrorText });
+        }
+      } else {
+        posthog.sendEvent(PostHogAction.SwapsBuildQuote, {
+          tokenIn: tokenB.name,
+          tokenOut: tokenA.name,
+          amount: quantity,
+          excludedDexs
+        });
+        const parsedResponse = (await response.json()) as BuildSwapResponse;
+        // Basic validation: ensure response has required fields
+        if (!parsedResponse || typeof parsedResponse.tx !== 'string' || typeof parsedResponse.p !== 'boolean') {
+          const errorMessage = 'Invalid build swap response structure';
+          logger.error(errorMessage, parsedResponse);
+          toast.notify({ duration: 3, text: unableToBuildErrorText });
+          return;
+        }
+        setBuildResponse(parsedResponse);
+        cb();
+      }
+    },
+    [addresses, tokenA, tokenB, quantity, targetSlippage, collateral, excludedDexs, utxos, t, posthog]
+  );
+
+  const signAndSubmitSwapRequest = useCallback(async () => {
+    const unableToSignErrorText = t('swaps.error.unableToSign');
+    if (!unsignedTx) {
+      toast.notify({ duration: 3, text: unableToSignErrorText });
+      posthog.sendEvent(PostHogAction.SwapsSignFailure);
+      setStage(SwapStage.Failure);
+      return;
+    }
+    try {
+      // Preserve the original transaction body and only update the witnesses
+      // This ensures the protocol parameters view hash remains correct
+      const unsignedTxFromCbor = Serialization.Transaction.fromCbor(unsignedTx.tx as unknown as Serialization.TxCBOR);
+      const finalTx = await inMemoryWallet.finalizeTx({
```

Comment:
```suggestion
      const {
          witness: { signatures }
        } = await inMemoryWallet.finalizeTx({
```

Maybe this would create less confusion